Data race in `inclusiveMetricsEval`

### Apache Iceberg version

None

### Please describe the bug 🐞

When running the `TestMergeManifests` test with `-race` flag, a data report is created (attached to the end of the message).
I only skimmed the code, but it looks like the culprit is concurrent fetching of manifest entries [here](https://github.com/apache/iceberg-go/blob/85f700808061fcea3655412245ef5d59f1784b69/table/scanner.go#L356). All concurrent requests write to the same structure [here](https://github.com/apache/iceberg-go/blob/85f700808061fcea3655412245ef5d59f1784b69/table/evaluators.go#L754-L756).

I confirmed that setting the concurrency here to `1` makes all the tests pass the data race check.

```
$ go test -race -run TestTableWriting/TestMergeManifests
==================
WARNING: DATA RACE
Write at 0x00c000f7dc20 by goroutine 353:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:754 +0xb9
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Previous write at 0x00c000f7dc20 by goroutine 354:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:754 +0xb9
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Goroutine 353 (running) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44

Goroutine 354 (finished) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44
==================
==================
WARNING: DATA RACE
Write at 0x00c000f7dc28 by goroutine 353:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:754 +0xf9
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Previous write at 0x00c000f7dc28 by goroutine 354:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:754 +0xf9
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Goroutine 353 (running) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44

Goroutine 354 (finished) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44
==================
==================
WARNING: DATA RACE
Write at 0x00c000f7dc30 by goroutine 353:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:755 +0x164
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Previous write at 0x00c000f7dc30 by goroutine 354:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:755 +0x164
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Goroutine 353 (running) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44

Goroutine 354 (finished) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44
==================
==================
WARNING: DATA RACE
Write at 0x00c000f7dc38 by goroutine 353:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:756 +0x1f0
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Previous write at 0x00c000f7dc38 by goroutine 354:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:756 +0x1f0
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Goroutine 353 (running) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44

Goroutine 354 (finished) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44
==================
==================
WARNING: DATA RACE
Write at 0x00c000f7dc40 by goroutine 353:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:756 +0x232
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Previous write at 0x00c000f7dc40 by goroutine 354:
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval()
      /dev/repo/iceberg-go/table/evaluators.go:756 +0x232
  github.com/apache/iceberg-go/table.(*inclusiveMetricsEval).Eval-fm()
      <autogenerated>:1 +0x47
  github.com/apache/iceberg-go/table.openManifest()
      /dev/repo/iceberg-go/table/scanner.go:142 +0x1b2
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries.func1()
      /dev/repo/iceberg-go/table/scanner.go:358 +0xd0
  golang.org/x/sync/errgroup.(*Group).add.func1()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x141

Goroutine 353 (running) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44

Goroutine 354 (finished) created at:
  golang.org/x/sync/errgroup.(*Group).add()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0xe4
  golang.org/x/sync/errgroup.(*Group).Go()
      /dev/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:93 +0x791
  github.com/apache/iceberg-go/table.(*Scan).collectManifestEntries()
      /dev/repo/iceberg-go/table/scanner.go:356 +0x531
  github.com/apache/iceberg-go/table.(*Scan).PlanFiles()
      /dev/repo/iceberg-go/table/scanner.go:399 +0xb1
  github.com/apache/iceberg-go/table.(*Scan).ToArrowRecords()
      /dev/repo/iceberg-go/table/scanner.go:441 +0x64
  github.com/apache/iceberg-go/table.(*Scan).ToArrowTable()
      /dev/repo/iceberg-go/table/scanner.go:474 +0x8c
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1125 +0x21f0
  github.com/apache/iceberg-go/table_test.arrowTableWithNull()
      /dev/repo/iceberg-go/table/table_test.go:974 +0x14b
  github.com/apache/iceberg-go/table_test.(*TableWritingTestSuite).TestMergeManifests()
      /dev/repo/iceberg-go/table/table_test.go:1067 +0x7d1
  runtime.call16()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/reflect/value.go:368 +0xb5
  github.com/stretchr/testify/suite.Run.func1()
      /dev/go/pkg/mod/github.com/stretchr/testify@v1.10.0/suite/suite.go:202 +0x6ed
  testing.tRunner()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /dev/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.2.linux-amd64/src/testing/testing.go:1851 +0x44
==================
--- FAIL: TestTableWriting (0.64s)
    --- FAIL: TestTableWriting/TestMergeManifests (0.37s)
        testing.go:1490: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/apache/iceberg-go/table	0.948s
```

--- a/table/evaluator_race_test.go
+++ b/table/evaluator_race_test.go
@@ -1,0 +1,88 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+
+package table
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/apache/iceberg-go"
+)
+
+// TestManifestEvalVisitorEvalRace fans out 256 goroutines that all call the
+// same cached Eval closure concurrently. Under -race, any write to a shared
+// receiver field inside Eval (e.g. m.partitionFields = parts) plus a read
+// inside VisitGreater/VisitLess/etc. will trip the detector.
+//
+// This reproduces the same shared-receiver pattern fixed previously in
+// exprEvaluator: newManifestEvaluator returns (&manifestEvalVisitor{...}).Eval,
+// callers in transaction.go's classifyFilesForFilteredDeletions loop the
+// returned closure across an errgroup of goroutines, and every goroutine
+// races on the captured *manifestEvalVisitor.
+func TestManifestEvalVisitorEvalRace(t *testing.T) {
+	intMin := []byte{0x00, 0x00, 0x00, 0x00}
+	intMax := []byte{0x7f, 0x00, 0x00, 0x00}
+
+	schema := iceberg.NewSchema(0,
+		iceberg.NestedField{
+			ID: 1, Name: "id",
+			Type: iceberg.PrimitiveTypes.Int32, Required: false,
+		},
+	)
+
+	spec := iceberg.NewPartitionSpec(iceberg.PartitionField{
+		Name:      "id",
+		SourceIDs: []int{1},
+		FieldID:   1000,
+		Transform: iceberg.IdentityTransform{},
+	})
+
+	// Single shared evaluator — this is the analogue of
+	// manifestEvaluators.Get(specID) returning the same cached closure to
+	// every goroutine in the production callers.
+	eval, err := newManifestEvaluator(spec, schema,
+		iceberg.LessThan(iceberg.Reference("id"), int32(50)),
+		true)
+	if err != nil {
+		t.Fatalf("newManifestEvaluator: %v", err)
+	}
+
+	const goroutines = 256
+
+	manifests := make([]iceberg.ManifestFile, goroutines)
+	for i := range manifests {
+		manifests[i] = iceberg.NewManifestFile(1, "", 0, 0, int64(i)).Partitions(
+			[]iceberg.FieldSummary{
+				{
+					ContainsNull: false,
+					LowerBound:   &intMin,
+					UpperBound:   &intMax,
+				},
+			},
+		).Build()
+	}
+
+	start := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(goroutines)
+	for i := 0; i < goroutines; i++ {
+		i := i
+		go func() {
+			defer wg.Done()
+			<-start
+			if _, err := eval(manifests[i]); err != nil {
+				t.Errorf("eval[%d]: %v", i, err)
+			}
+		}()
+	}
+	close(start)
+	wg.Wait()
+}

--- a/table/evaluators.go
+++ b/table/evaluators.go
@@ -60,13 +60,21 @@ type manifestEvalVisitor struct {
 }
 
 func (m *manifestEvalVisitor) Eval(manifest iceberg.ManifestFile) (bool, error) {
-	if parts := manifest.Partitions(); len(parts) > 0 {
-		m.partitionFields = parts
-
-		return iceberg.VisitExpr(m.partitionFilter, m)
+	parts := manifest.Partitions()
+	if len(parts) == 0 {
+		return rowsMightMatch, nil
 	}
 
-	return rowsMightMatch, nil
+	// Use a per-call visitor so concurrent callers of the same cached Eval
+	// closure (e.g. the errgroup in classifyFilesForFilteredDeletions and
+	// the parallel manifest scans in scanner.collectManifestEntries) do
+	// not race on partitionFields. Mirrors the exprEvaluator fix above.
+	ev := manifestEvalVisitor{
+		partitionFields: parts,
+		partitionFilter: m.partitionFilter,
+	}
+
+	return iceberg.VisitExpr(ev.partitionFilter, &ev)
 }
 
 func removeBoundCmp[T iceberg.LiteralType](bound iceberg.Literal, vals []iceberg.Literal, cmpToDelete int) []iceberg.Literal {

--- a/visitors.go
+++ b/visitors.go
@@ -201,9 +201,13 @@ type exprEvaluator struct {
 }
 
 func (e *exprEvaluator) Eval(st StructLike) (bool, error) {
-	e.st = st
+	// Use a per-call evaluator so concurrent callers (e.g. parallel
+	// manifest scans in table.Scan.collectManifestEntries) do not race
+	// on the shared `st` field. Mirrors the fix in inclusiveMetricsEval
+	// from #445.
+	ev := exprEvaluator{bound: e.bound, st: st}
 
-	return VisitExpr(e.bound, e)
+	return VisitExpr(ev.bound, &ev)
 }
 
 func (e *exprEvaluator) VisitUnbound(UnboundPredicate) bool {

